crates/factor-outbound-pg/tests/factor_test.rs

```diff
@@ -0,0 +1,142 @@
+use spin_factors::{anyhow, RuntimeFactors};
+use spin_factors_test::{toml, TestEnvironment};
+use spin_world::async_trait;
+use spin_world::v2::postgres::HostConnection;
+use spin_world::v2::rdbms_types::Error as PgError;
+use tokio_postgres::types::ToSql;
+use tokio_postgres::{Error, Row, ToStatement};
+
+#[derive(RuntimeFactors)]
+struct TestFactors {
+    wasi: WasiFactor,
+    variables: VariablesFactor,
+    networking: OutboundNetworkingFactor,
+    pg: OutboundPgFactor<MockClient>,
+}
+
+fn factors() -> Result<TestFactors> {
+    let mut f = TestFactors {
+        wasi: WasiFactor::new(DummyFilesMounter),
+        variables: VariablesFactor::default(),
+        networking: OutboundNetworkingFactor,
+        pg: OutboundPgFactor::<MockClient>::default(),
+    };
+    f.variables.add_provider_type(StaticVariables)?;
+    Ok(f)
+}
+
+fn test_env() -> TestEnvironment {
+    TestEnvironment::default_manifest_extend(toml! {
+        [component.test-component]
+        source = "does-not-exist.wasm"
+        allowed_outbound_hosts = ["postgres://*:*"]
+    })
+}
+
+#[tokio::test]
+async fn disallowed_host_fails() -> anyhow::Result<()> {
+    let factors = factors()?;
+    let env = TestEnvironment::default_manifest_extend(toml! {
+        [component.test-component]
+        source = "does-not-exist.wasm"
+    });
+    let mut state = env.build_instance_state(factors).await?;
+
+    let res = state
+        .pg
+        .open("postgres://postgres.test:5432/test".to_string())
+        .await;
+    let Err(err) = res else {
+        bail!("expected Err, got Ok");
+    };
+    assert!(matches!(err, PgError::ConnectionFailed(_)));
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn allowed_host_succeeds() -> anyhow::Result<()> {
+    let factors = factors()?;
+    let env = test_env();
+    let mut state = env.build_instance_state(factors).await?;
+
+    let res = state
+        .pg
+        .open("postgres://localhost:5432/test".to_string())
+        .await;
+    let Ok(_) = res else {
+        bail!("expected Ok, got Err");
+    };
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn exercise_execute() -> anyhow::Result<()> {
+    let factors = factors()?;
+    let env = test_env();
+    let mut state = env.build_instance_state(factors).await?;
+
+    let connection = state
+        .pg
+        .open("postgres://localhost:5432/test".to_string())
+        .await?;
+
+    state
+        .pg
+        .execute(connection, "SELECT * FROM test".to_string(), vec![])
+        .await?;
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn exercise_query() -> anyhow::Result<()> {
+    let factors = factors()?;
+    let env = test_env();
+    let mut state = env.build_instance_state(factors).await?;
+
+    let connection = state
+        .pg
+        .open("postgres://localhost:5432/test".to_string())
+        .await?;
+
+    state
+        .pg
+        .query(connection, "SELECT * FROM test".to_string(), vec![])
+        .await?;
+
+    Ok(())
+}
+
+pub struct MockClient {}
```

Comment:
Is the plan to eventually expand this mock implementation to track how it's been called and let the tests assert that the implementation has been called properly?

Reply:
Yep